store/gaskvstore.go

```diff
@@ -120,36 +127,42 @@ func newGasIterator(gasMeter sdk.GasMeter, gasConfig sdk.GasConfig, parent sdk.I
 }
 
 // Implements Iterator.
-func (g *gasIterator) Domain() (start []byte, end []byte) {
-	return g.parent.Domain()
+func (gi *gasIterator) Domain() (start []byte, end []byte) {
+	return gi.parent.Domain()
 }
 
 // Implements Iterator.
-func (g *gasIterator) Valid() bool {
-	return g.parent.Valid()
+func (gi *gasIterator) Valid() bool {
+	return gi.parent.Valid()
 }
 
-// Implements Iterator.
-func (g *gasIterator) Next() {
-	g.parent.Next()
+// Next implements the Iterator interface. It iterates to the next key/value
+// pair in the iterator. It incurs a flat gas cost for iteration and a variable
+// gas cost based on the current value's length.
+func (gi *gasIterator) Next() {
+	value := gi.Value()
+
+	gi.gasMeter.ConsumeGas(gi.gasConfig.ValueCostPerByte*sdk.Gas(len(value)), sdk.GasValuePerByteDesc)
```

Comment:
Is there a measurable difference in time for reading a few bytes, vs reading a lot of bytes? This feels a bit like a micro-charge, in the context of the things which will be charging for.

Reply:
@ValarDragon I have no performed any benchmarks, but I'd imagine there would be some noticeable difference when reading very large values. Would you like me to include some benchmarks?

Reply:
Although the cost relative to the overhead of just reading any value at all is low, I think we should definitely do this - (a) it provides additional protection against large user-generated values (e.g. memos) and (b) it provides some (bad) approximation of the storage cost, which is helpful since we don't have any concept of storage rent (yet).

Reply:
Agreed!

Reply:
I'd prefer we make the gas cost of this negligible until we get benchmarks. However currently we do gas as an int64, so I'm not sure if there is a "negligible value". Anything we charge gas on now is something we'll have to benchmark / price sanely. Benchmarking IAVL on different read sizes is probably going to end up being quite hard, since there are many different parameters, and its going to vary _wildly_ across systems and databases.

> it provides additional protection against large user-generated values (e.g. memos) and

Memos are never read from state on-chain. Additionally we have open issue to prune those after a short number of blocks. What are other large user-generated values where this is applicable? I don't think bad approximations are that helpful, lets just charge gas for tx size. 

This seems like a premature thing to put in until we get concrete benchmarks. Given that we're probably going to have to have a large gas update _anyway_, imo we should hold off on this particular point / make it charge a negligible fraction.

Reply:
Other user generated values: validator descriptions, several governance proposal fields, IBC packet fields. We should definitely also charge gas for tx size, but it should cost more to write to the IAVL tree, don't you think?